components/eamxx/src/share/field/field.hpp

```diff
@@ -316,16 +328,42 @@ class Field {
   void sync_views_impl () const;
 
   template<HostOrDevice HD, typename ST>
-  void deep_copy_impl (const ST value) const;
+  void deep_copy_impl (const ST value);
 
-  template<HostOrDevice HD, typename ST>
-  void deep_copy_impl (const Field& src) const;
-
-  template<CombineMode CM, HostOrDevice HD, typename ST>
-  void update_impl (const Field& x, const ST alpha, const ST beta, const ST fill_val);
+  // The update method calls this, with ST matching this field data type.
+  // Note: use_fill is used to determine *at compile time* whether to use
+  // the combine<CM> utility or combine_and_fill<CM>
+  template<CombineMode CM, HostOrDevice HD, bool use_fill, typename ST, typename XST>
+  void update_impl (const Field& x, const ST alpha, const ST beta, const ST fill_val = 0);
 
 protected:
 
+  // This is helpful in the update method, and it can't be a lambda due to templating
+  // NOTE: if neither lsh nor rhs has "mask_value" extra data, this will throw.
+  template<typename T>
+  static T get_mask_value (const Field& lhs, const Field& rhs) {
+    // If present, prefer using the lhs native mask value.
+    if (lhs.get_header().has_extra_data("mask_value")) {
+      return lhs.get_header().get_extra_data<T>("mask_value");
+    } else if (rhs.get_header().has_extra_data("mask_value")) {
+      return rhs.get_header().get_extra_data<T>("mask_value");
+    } else {
+      EKAT_ERROR_MSG ("Error! Neither of the fields has the 'mask_value' extra data.\n"
+          " - lhs name: " + lhs.name() + "\n"
+          " - rhs name: " + rhs.name() + "\n");
+
+      return 0; // unreachable return
+    }
+  }
+
+  template<int N,HostOrDevice HD>
+  using MDRange = Kokkos::MDRangePolicy<typename get_device<HD>::execution_space,
```

Comment:
Seems like you are redefining `MDRange` in the `scream::impl::CombineViewsHelper` below instead of using this one?

Reply:
Ah, yes. I meant to rm this one when I created that helper struct. I will prune and push.